File: 638. Shopping Offers/C++/Solution.cpp
class Solution {
public:
    int shoppingOffers(vector<int> &price, vector<vector<int>> &special, vector<int> &needs) {
        class Utils {
        public:
            string toKey(vector<int> &array) {
                string builder = "";
                for (int item : array) {
                    builder += to_string(item);
                }
                return builder;
            }

            vector<int> toArray(string key) {
                vector<int> array;
                for (auto letter : key) {
                    array.push_back(letter - '0');
                }
                return array;
            }

            int trace(vector<int> &prices, vector<vector<int>> &specials, string needsKey,
                      unordered_map<string, int> &memory) {
                if (memory.count(needsKey)) {
                    return memory[needsKey];
                }
                vector<int> needsArray = toArray(needsKey);
                int result = 0;
                for (int i = 0; i < prices.size(); i++) {
                    result += prices[i] * needsArray[i];
                }
                for (vector<int> &special : specials) {
                    vector<int> newNeedsArray = needsArray;
                    bool canFit = true;
                    for (int i = 0; i < needsArray.size(); i++) {
                        if (needsArray[i] < special[i]) {
                            canFit = false;
                            break;
                        }
                        newNeedsArray[i] -= special[i];
                    }
                    if (canFit) {
                        result = min(result, special.back() + trace(prices, specials, toKey(newNeedsArray), memory));
                    }
                }
                memory[needsKey] = result;
                return result;
            }
        };
        Utils utils;
        unordered_map<string, int> memory;
        return utils.trace(price, special, utils.toKey(needs), memory);
    }
};

File: 638. Shopping Offers/Python/Solution.py
class Solution:
    def shoppingOffers(self, price, special, needs):
        """
        :type price: List[int]
        :type special: List[List[int]]
        :type needs: List[int]
        :rtype: int
        """

        def to_key(array):
            key = ''
            for item in array:
                key += str(item)
            return key

        def to_array(key):
            array = []
            for letter in key:
                array.append(int(letter))
            return array

        def trace(prices, specials, needs_key, memory):
            if needs_key in memory:
                return memory[needs_key]
            needs_array = to_array(needs_key)
            result = 0
            for price, need in zip(prices, needs_array):
                result += price * need
            for special in specials:
                new_needs_array = needs_array[:]
                can_fit = True
                for i in range(len(needs_array)):
                    if needs_array[i] < special[i]:
                        can_fit = False
                        break
                    new_needs_array[i] -= special[i]
                if can_fit:
                    result = min(result, special[-1] + trace(prices,
                                                             specials, to_key(new_needs_array), memory))
            memory[needs_key] = result
            return result

        return trace(price, special, to_key(needs), {})
